### PR TITLE
Refactor services and Azure clients to make services unit testable

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**:
As I was working on #340, it became obvious that the services/ packages have no unit test coverage. I started this branch in order to add tests but quickly realized that the current state of things made it quite difficult: 
1)  Mocking Azure go-autorest futures is not possible which meant I couldn't just use the mocks from Azure sdk-for-go computeapi/networkapi/resourceapi (https://github.com/Azure/go-autorest/issues/469)
2) The `.NewService()` calls within Service functions (eg. https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/cloud/services/virtualmachines/virtualmachines.go#L94) make it impossible to unit test said functions because, even when using a fake Azure Client for the Service resource type, the test will attempt to create a new client from the fake sub ID and authorizer which will fail. 
3) There is a missing function in the Azure sdk-for-go ResourceSkusClientAPI (opened https://github.com/Azure/azure-sdk-for-go/issues/6559)

I was able to find workarounds for 2 and 3 in my first attempt, but 1 was the real blocker here. In order to fix it, I added a new `Client` interface for each Azure resource service and then extracted the calls to the Azure SDK from the functions that contain Get/Reconcile/Delete logic (which also happens to be an overall improvement). The new `Client` interfaces are now completely mockable and the mocks are generated using GoMock.

I also added a `doc.go` to each mock service package (following what capa had done) to auto generate the mocks whenever `go generate` is run.

In the interest of keeping this PR from being completely un-reviewable (it's already XXL as-is, although a lot of the files are auto-generated), I only added a simple unit test for `virtualmachines.go`'s `Reconcile()` to demonstrate how the tests will work. I would also like to get reviews on the overall approach before spending too much time writing tests for each package. I plan on getting this PR merged first, then rebasing #340 and adding vnet tests, and finally adding unit tests for the rest of the services packages.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

I will rebase the PR before it merges, keeping commit history for now to keep trace of different approaches tried.

/hold for rebase

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by devigned** [comment]
Ouch... That was a big refactor, but definitely one that makes it much easier to test. Looks awesome!

File: cloud/services/routetables/routetables.go [hunk 1]
```diff
@@ -86,7 +77,7 @@ func (s *Service) Delete(ctx context.Context, spec interface{}) error {
 		return errors.New("Invalid Route Table Specification")
 	}
 	klog.V(2).Infof("deleting route table %s", routeTableSpec.Name)
-	future, err := s.Client.Delete(ctx, s.Scope.AzureCluster.Spec.ResourceGroup, routeTableSpec.Name)
+	err := s.Client.Delete(ctx, s.Scope.AzureCluster.Spec.ResourceGroup, routeTableSpec.Name)
```
Comment (devigned) [inline cloud/services/routetables/routetables.go]: This looks so much cleaner than exposing the future upward from the go-sdk impl.

**Comment by devigned** [comment]
/hold

`cloud/services/securitygroups/service.go:35: File is not `gofmt`-ed with `-s` (gofmt)`

**Comment by CecileRobertMichon** [comment]
Rebase done.

/hold cancel

**Comment by justaugustus** [comment]
Awesome job again, @CecileRobertMichon! Thanks for the commit cleanup. :)
/lgtm
/approve

**Pull request merged**